### PR TITLE
Bytes type and coverall test fixture

### PR DESCRIPTION
**Pull request by skeet70**
Switched from `Disposable` to `AutoClosable`. Big difference is that `AutoClosable` `throws Exception` which makes it slightly more annoying in some use cases. It allows us to use `try-with-resources` though to make sure things that need to get destroyed do get destroyed. We can't inline functions like Kotlin is doing with `destroy` for `Disposable`, so we have to hold a strong reference to be able to `destroy` something after a block of code, which negates a bit of the point.

Disposable could be added back in the future (or in addition to AutoClosable) if necessary, but looking through the Rust code I didn't see anywhere a `destroy` was being presented from that side or anything.

Writing the coverall tests did catch a few edge cases that were fixed in this PR.

### CODE REVIEW

File: src/templates/ByteArrayHelper.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    public byte[] read(ByteBuffer buf) {
+        int len = buf.getInt();
+        byte[] byteArr = new byte[len];
+        buf.get(byteArr);
+        return byteArr;
+    }
+  
+    @Override
+    public long allocationSize(byte[] value) {
+        return 4L + (long)value.length;
```
Comment (BobWall23) [inline src/templates/ByteArrayHelper.java:18]: length always 32 bits, even on 64 bit architectures? Looks like Java doesn't have `sizeof()`, so doing something different would probably suck anyway.

File: src/templates/ObjectTemplate.java [hunk 2]
```diff
@@ -151,15 +152,15 @@ public class {{ impl_class_name }} implements Disposable, AutoCloseable, {{ inte
...
+  public {{ impl_class_name }}({% call java::arg_list(cons, true) -%}) {% match cons.throws_type() %}{% when Some(throwable) %}throws {{ throwable|type_name(ci) }}{% else %}{% endmatch %}{
     this((Pointer){%- call java::to_ffi_call(cons) -%});
   }
   {%-     endif %}
   {%- when None %}
   {%- endmatch %}
 
   @Override
-  public void destroy() {
+  public synchronized void close() {
```
Comment (BobWall23) [inline src/templates/ObjectTemplate.java:163]: Since this is `close()` instead of `destroy()`, should it be `wasClosed` instead of `wasDestroyed`? Maybe `wasDestroyed` is fine, since it looks like it does destroy the fields when closing.

File: src/templates/RustBufferTemplate.java [hunk 3]
```diff
@@ -69,6 +75,19 @@ public void setValue(RustBuffer.ByValue value) {
...
         pointer.setPointer(8, value.data);
     }
+
+    /**
+     * Get a `RustBuffer.ByValue` from this reference.
+     */
+    public RustBuffer.ByValue getValue() {
+        Pointer pointer = this.getPointer();
+        RustBuffer.ByValue value = new RustBuffer.ByValue();
+        value.writeField("capacity", pointer.getLong(0));
```
Comment (BobWall23) [inline src/templates/RustBufferTemplate.java:85]: And the Rust ones are always 64 bits for capacity and length, regardless of architecture?

**Review by BobWall23 — APPROVED** [review]

**Review by coltfred — APPROVED** [review]

File: src/templates/Helpers.java [hunk 4]
```diff
@@ -125,22 +131,23 @@ static <T> void uniffiTraitInterfaceCall(
       try {
           writeReturn.accept(makeCall.get());
       } catch (Exception e) {
+          System.out.println(e);
```
Comment (coltfred) [inline src/templates/Helpers.java:134]: rogue println?

File: src/templates/ByteArrayHelper.java (hunk 1 above)
Comment (skeet70) [reply src/templates/ByteArrayHelper.java:18]: The return value from Java for `length` is always an `int` regardless of arch/JVM. The actual max size of the array is JVM/arch dependant, but pretty sure it's always within int max.

File: src/templates/ObjectTemplate.java (hunk 2 above)
Comment (skeet70) [reply src/templates/ObjectTemplate.java:163]: Yeah it's a bit of both. `close` is the Java parlance and the Java interface we've implemented, but `destroy` is the common-across-bindgens uniffi word for it.

File: src/templates/RustBufferTemplate.java (hunk 3 above)
Comment (skeet70) [reply src/templates/RustBufferTemplate.java:85]: This is a construct just for usage with uniffi, so yeah the capacity and length is always 64 bit.
